CTF101 crypto writeups (#14)
stellarvector/blog: _writeups/2023/ctf101/rsa-challenges.md

---

> layout: writeup

Reviewer: ```suggestion
```
`mathjax` is based on `writeup` (otherwise the layouts would be broken)

print(f'{m = }')
# m = b'sv{RSA_1s_fun}'
```

## SPQR

The second challenge was the same, except that this time we are given `n = 101264313152893175913493633998971546542632990679253512564207556023
` instead of $$p$$ and $$q$$. To recover them we have to factorize. The number is too big to use too simple methods like trial division, but can be easily handled by software like `sagemath`, `MAGMA` or the online [Alpertron Factorization Tool](https://www.alpertron.com.ar/ECM.HTM). Once we recover $$p$$ and $$q$$, we can apply the same decryption procedure as above and get the flag: `sv{f4ct0r1ng_1s_funn13r}`.

## leak

The final challenge was meant to be an actual RSA challenge. First of all, `n` is a (kind of) RSA size modulus:

```python
m = bytes_to_long(flag)
e = 65537
p = getPrime(512)
q = getPrime(512)
n = p*q
```

Factoring a `1024`-bit number directly is hopeless, so together with `c` we are given some *leaks*:

```python
phi = (p-1)*(q-1)
assert GCD(e, phi) == 1
c = pow(m, e, n)
print(f'{c = }')

d = pow(e, -1, phi)
leak = d*e - 1
for i in range(2):
	k = randint(21, 42)
	leak_i = leak * getPrime(512) + k
	print(f'leak_{i+1} = {leak_i}')
```

> We are given two leaks: the value `leak = d*e - 1` is multiplied by two random primes of `512` bits each, then a little error `k` (a random number between `21` and `42`) is added to each one, and then they are returned to us. The firts step here is to recover the value of `leak`. Then we have to figure out how to factor exploiting this extra information (the challenge had a hint suggesting that this is possible).

Reviewer: ```suggestion
We are given two leaks: the value `leak = d*e - 1` is multiplied by two random primes of `512` bits each, then a little error `k` (a random number between `21` and `42`) is added to each one, and then they are returned to us. The first step here is to recover the value of `leak`. Then we have to figure out how to factor the number, exploiting this extra information (the challenge had a hint suggesting that this is possible).
```

So let's recover the `leak`. The first observation here is that if we knew `k1` and `k2` (the two values of `k` respectively), then we could recover `leak` by taking the `GCD(leak_1, leak2)`. This is true since each of them is multiplied by a big prime, which is a factor they are not likely to share. The same observation allows us to bruteforce `k1` and `k2`: we have in total ~400 possible pairs `(k1, k2)` to check, and for each of those we can just check if we get a reasonable `GCD`. If `k1` or `k2` are wrong, the resulting `leak_i - ki` will not share the common factor `leak`, and hence the `GCD` we get would probably be small.

```python
phi = (p-1)*(q-1)

assert GCD(e, phi) == 1

c = pow(m, e, n)
print(f'{c = }')

d = pow(e, -1, phi)
leak = d*e - 1

for i in range(2):
	k = randint(21, 42)

	leak_i = leak * getPrime(512) + k
	print(f'leak_{i+1} = {leak_i}')
```

> Running this we get `k1 = 25` and `k2 = 27`, but most importantly the value of `leak`. Now the second part: how do we factor knowing `leak = d*e - 1`? As we can se from the challenge script (or from a generic knowledge of how RSA works), we know that it holds

Reviewer: ```suggestion
Running this we get `k1 = 25` and `k2 = 27`, but most importantly the value of `leak`. Now the second part: how do we factor knowing `leak = d*e - 1`? As we can see from the challenge script (or from a generic knowledge of how RSA works), we know that the following holds:
```
